lib/qs/worker.rb

```diff
@@ -0,0 +1,64 @@
+require 'dat-worker-pool/worker'
+
+module Qs
+
+  module Worker
+
+    IVAR_NAME = "@qs_worker_mixin_included_already".freeze
+
+    def self.included(klass)
+      return if klass.instance_variable_get(IVAR_NAME)
+
+      klass.class_eval do
+        include DatWorkerPool::Worker
+        include InstanceMethods
+
+        on_available{ params[:qs_worker_available].signal }
+
+        on_error{ |e, wi| qs_handle_exception(e, wi) }
+
+      end
+
+      klass.instance_variable_set(IVAR_NAME, true)
+    end
+
+    module InstanceMethods
+
+      def work!(queue_item)
+        Qs::PayloadHandler.new(params[:qs_daemon_data], queue_item).run
+      end
+
+      private
+
+      # * This only catches errors that happen outside of running the payload
+      #   handler. The only known use-case for this is dat worker pools
+      #   hard-shutdown errors.
+      # * If there isn't a queue item (this can happen when an idle worker is
+      #   being forced to exit) then we don't need to do anything.
+      # * If we never started processing the queue item, its safe to requeue it.
+      #   Otherwise it happened while processing so the payload handler caught
+      #   it or it happened after the payload handler which we don't care about.
```

Comment:
@jcredding we've gone away from this "bulleted" comment style - can we clean this up?